src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs

```diff
@@ -487,51 +487,216 @@ void checkConstraintLanguageVersionAndRuntimeSupportForConversion(SyntaxNode syn
+            Func<TArg, TypeSymbol, bool>? returnTypePredicate = null)
+        {
+            return TryFindSingleMember(type, WellKnownMemberNames.ImplicitConversionName, (parameterPredicate, returnTypePredicate, arg),
+                static (arg, method) => method is
                 {
                     ParameterCount: 1,
                     Arity: 0,
                     IsStatic: true,
                     DeclaredAccessibility: Accessibility.Public,
-                    Parameters: [{ Type: ArrayTypeSymbol { IsSZArray: true, ElementType: TypeParameterSymbol } }]
-                });
+                    Parameters: [{ Type: { } parameterType }],
+                } && arg.parameterPredicate(arg.arg, parameterType) &&
+                    arg.returnTypePredicate?.Invoke(arg.arg, method.ReturnType) != false);
+        }
+
+        internal static bool NeedsSpanCastUp(TypeSymbol source, TypeSymbol destination)
+        {
+            Debug.Assert(source.OriginalDefinition.IsSpan() || source.OriginalDefinition.IsReadOnlySpan());
+            Debug.Assert(destination.OriginalDefinition.IsReadOnlySpan());
+            Debug.Assert(!source.IsDefinition && !destination.IsDefinition);
+
+            var sourceElementType = ((NamedTypeSymbol)source).TypeArgumentsWithAnnotationsNoUseSiteDiagnostics[0].Type;
+            var destinationElementType = ((NamedTypeSymbol)destination).TypeArgumentsWithAnnotationsNoUseSiteDiagnostics[0].Type;
+
+            var sameElementTypes = sourceElementType.Equals(destinationElementType, TypeCompareKind.AllIgnoreOptions);
+
+            Debug.Assert(!source.OriginalDefinition.IsReadOnlySpan() || !sameElementTypes);
+
+            return !sameElementTypes;
+        }
+
+        // ReadOnlySpan<T> ReadOnlySpan<T>.CastUp<TDerived>(ReadOnlySpan<TDerived>) where TDerived : class
+        internal static MethodSymbol? TryFindCastUpMethod(TypeSymbol source, TypeSymbol destination)
+        {
+            Debug.Assert(source.IsReadOnlySpan() && destination.IsReadOnlySpan());
+            Debug.Assert(source.IsDefinition && destination.IsDefinition);
+
+            return TryFindSingleMember(destination, WellKnownMemberNames.CastUpMethodName, (source, destination),
+                static (arg, method) => method is
+                {
+                    ParameterCount: 1,
+                    Arity: 1,
+                    IsStatic: true,
+                    DeclaredAccessibility: Accessibility.Public,
+                    Parameters: [{ } parameter],
+                    TypeArgumentsWithAnnotations: [{ } typeArgument],
+                } &&
+                    // parameter type is the source ReadOnlySpan<>
+                    arg.source.Equals(parameter.Type.OriginalDefinition, TypeCompareKind.ConsiderEverything) &&
+                    // return type is the destination ReadOnlySpan<>
+                    arg.destination.Equals(method.ReturnType.OriginalDefinition, TypeCompareKind.ConsiderEverything) &&
```

Comment:
It looks like these two checks are just checking that the parameter type and return type are some constructed instances of `ReadOnlySpan<T>` since `arg.source` and `arg.destination` are both definitions.

If that's the case, can we simplify those checks to `parameter.Type.IsReadOnlySpan() && method.ReturnType.IsReadOnlySpan()`?

And if so, do we need to pass in `(source, destination)` as the `arg` for `TryFindSingleMember()` or could we just pass `destination.TypeParameters[0]` instead?

Reply:
This check also makes sure the ReadOnlySpan type is not different than the one used in the conversion. For example test `Conversion_ReadOnlySpan_ReadOnlySpan_Implicit_MultipleSpans_02` (and _03, _04) would behave differently if we checked just `IsReadOnlySpan()`.